Catch unbound variables mini

cataldor opened the pull request:
Make mini versions compatible with scripts using the -u shell flag by
capturing unset variables.

This commit makes _check_vars accept empty parameters.

---

This patch aims to allow users to use `-u` flag in their script while calling the mini version.

I'm not sure how you are suppose to create the array in the mini version (also, the mini versions are not in the test scripts as far as i know), so i created a "bastard" version of the declare functions to test it.

Before this patch series:
```console
./posix-arrays-i-mini.sh: line 70: $2: unbound variable
./posix-arrays-a-mini.sh: line 69: $3: unbound variable
```

After this patch series:
```console
0 1 2 3 4 5 6 7 8
999
declare_i_arr: Error: invalid array name 'iarr_*'.
set_i_arr_el: Error: invalid array name 'iarr_*'.
foo woo
foo,bar
```

Here is the script i used to test it:

indexed array test
```sh
#!/bin/sh

set -u

. ./posix-arrays-i-mini.sh

declare_i_arr() {
	___me="declare_i_arr"
	case "$*" in '') wrongargs "$@"; return 1; esac
	_arr_name="$1"; shift
	_check_vars "$_arr_name" || return 1

	_index=0; _indices=''
	for ___val in "$@"; do
		_indices="$_indices$_nl$_index"
		_index=$((_index + 1))
	done
	_index=$((_index - 1))

	case "$_index" in
		"-1") ;;
		*) eval "_i_${_arr_name}_h_index"='$_index'"
				_i_${_arr_name}_indices"='$_indices'"
				_i_${_arr_name}_sorted_flag=1"
	esac
	:
}


declare_i_arr iarr 0 1 2 3 4 5 6 7 8
get_i_arr_indices iarr retval
echo $retval
set_i_arr_el iarr 4 999
get_i_arr_val iarr 4 retval
echo $retval

declare_i_arr iarr_* 0 0 1
if [ $? -ne 1 ]; then
	echo "expected failure"
fi
set_i_arr_el iarr_* 4 999
if [ $? -ne 1 ]; then
	echo "expected failure"
fi
```

associative test
```sh
#!/bin/sh

set -eu

. ./posix-arrays-a-mini.sh

check_pair() { case "$___pair" in *=* ) ;; *) echo "$___me: Error: '$___pair' is not a 'key=value' pair." >&2; return 1; esac; }

declare_a_arr() {
	___me="declare_a_arr"
	case "$*" in '') wrongargs "$@"; return 1; esac
	_arr_name="$1"; shift
	_out_var=""
	___keys=''
	_check_vars "$_arr_name" "$_out_var" || return 1

	case "$*" in *?* )
		for ___pair in "$@"; do
			check_pair || return 1
			___key="${___pair%%=*}"
			___val="${___pair#*=}"
			_check_vars "$___key" "$_out_var"  "$_arr_name" || return 1 # todo: revert assignments
			eval "_a_${_arr_name}_${___key}"='$___val'
			___keys="$___keys$_nl$___key"
		done
	esac

	eval "_a_${_arr_name}___keys=\"$___keys\"; _a_${_arr_name}_sorted_flag=0"

	:
}

declare_a_arr aarr foo="bar" woo="zee"
get_a_arr_keys aarr retval
echo $retval
set_a_arr_el aarr "foo=foo,bar" 
get_a_arr_val aarr foo retval
echo $retval
```

rodrigo-cadore-la committed:
fix {___nl => _nl} for mini posix array versions

The ___nl variable was being referred as _nl in a number of function
calls. Rename the variable to _nl for matching them.

Signed-off-by: Rodrigo Cataldo <rodrigo.cadore@l-acoustics.com>

rodrigo-cadore-la committed:
Make mini versions -u compatible by catching unset variables

Make mini versions compatible with scripts using the -u shell flag by
capturing unset variables.

This commit makes _check_vars accept empty parameters.

Signed-off-by: Rodrigo Cataldo <rodrigo.cadore@l-acoustics.com>

friendly-bits commented:
Your suggestion for the _check_vars() modification should work, however this is a shorter version for the -a-mini script which also catches some additional possible issues when using the -u flag:

```
_check_vars() {
	case "${1:-}${2:-}${3:-}" in ''|*[!A-Za-z0-9_]* )
		for _test_seq in "_arr_name|array name" "___key|key" "_out_var|output variable name"; do
			eval "_var_val=\"\${${_test_seq%%|*}:-}\""; _var_desc="${_test_seq#*|}"
			case "$_var_val" in ''|*[!A-Za-z0-9_]* )
				printf '%s\n' "$___me: Error: invalid $_var_desc '$_var_val'." >&2
				return 1
			esac
		done
	esac
}

```

That said, I did not consider using the -u flag when writing this code, so you may get errors in multiple additional places, for example in set_a_arr_el():

`_arr_name="$1"; ___pair="$2"`

will produce a similar error if calling the function without the required arguments.

The point of the -u flag (as I understand it) is to prevent use of unset variables. My code performs the necessary checks in order to avoid this in the first place, for example in set_a_arr_el():

`case "$#" in 2) ;; *) wrongargs "$@"; return 1; esac`

In this instance the check comes after variables assignment, so while the function will produce the expected error without the -u flag, with that flag you will get a "parameter not set" error.

cataldor commented:
Hello friendly-bits, thanks for the reply. My comments are inline:

> Your suggestion for the _check_vars() modification should work, however this is a shorter version for the -a-mini script which also catches some additional possible issues when using the -u flag:
> 
> ```
> _check_vars() {
> 	case "${1:-}${2:-}${3:-}" in ''|*[!A-Za-z0-9_]* )
> 		for _test_seq in "_arr_name|array name" "___key|key" "_out_var|output variable name"; do
> 			eval "_var_val=\"\${${_test_seq%%|*}:-}\""; _var_desc="${_test_seq#*|}"
> 			case "$_var_val" in ''|*[!A-Za-z0-9_]* )
> 				printf '%s\n' "$___me: Error: invalid $_var_desc '$_var_val'." >&2
> 				return 1
> 			esac
> 		done
> 	esac
> }
> ```
> 
> A similar modification should do the job for the -i-mini script.

Nice!

> 
> That said, I did not consider using the -u flag when writing this code, so you may get errors in multiple additional places, for example in set_a_arr_el():
> 
> `_arr_name="$1"; ___pair="$2"`
> 
> will produce a similar error if calling the function without the required arguments.

Interesting I did not notice that.

> 
> The point of the -u flag (as I understand it) is to prevent use of unset variables. My code performs the necessary checks in order to avoid this in the first place, for example in set_a_arr_el():
> 
> `case "$#" in 2) ;; *) wrongargs "$@"; return 1; esac`

Well, i guess yes, but it is not the reason why I use it -- I typically use `-u` flag to catch typo mistakes (as was the case of the previous pull request). It happens with me quite frequently sadly.

So, the reasoning of sending this PR was not really to cover all cases (although, that would be ideal), but to make sure the 'happy path' (when the user respects the interface) would not hit a unset variable.

But I totally understand that you may not want to cover this use case, I will talk more about this at the end.

> 
> In this instance the check comes after variables assignment, so while the function will produce the expected error without the -u flag, with that flag you will get a "parameter not set" error.

True. Typically, i also use `-e` flag, so, in either case, the script will stop.

> 
> As to the test scripts, the mini versions are an afterthought and when splitting them out of the main script, I was too busy or too lazy (or both) to write more tests specifically for them. That said, my assumption is that since the main script is pretty well tested, the minified versions should be fine (and I'm using a slightly modified version of the -a-mini script in one of my other projects). The reason for the '_nl -- ___nl' issue is that I made some improvements in the downstream version which doesn't need the separate '_nl=[]' line and while subsequently updating the code upstream, I forgot to change that variable.

I understand.

My question for you is this: would you like me to cover all cases of unset variables in the mini-versions so we can merge it? or do you believe it is not worth it (also, I do not have the time to do the same for the full versions, so it is a bit inconsistent)

Personally, I do think there is value in avoiding the unset variables in the 'happy path'. So that a user (like me) can use it fine if it respects the interface.

friendly-bits commented:
> My question for you is this: would you like me to cover all cases of unset variables in the mini-versions so we can merge it?

I'll review and update the code for this purpose and then if you think additional changes will be needed, I'll welcome your input.

cataldor commented:
> > My question for you is this: would you like me to cover all cases of unset variables in the mini-versions so we can merge it?
> 
> I'll review and update the code for this purpose and then if you think additional changes will be needed, I'll welcome more input.

Nice! thank you

friendly-bits commented:
I updated the code for better compatibility with the -u shell option. The `_check_vars()` function now performs a more thorough check of the arguments, which results in degraded performance, especially in the faster functions (`set_[x]_arr_el()` and `get_[x]_arr_val()`). To mitigate that, I implemented support for the `nocheckvars` environment variable which forces skipping these checks. It may be used in production after the code is well tested.

You can still get an `unbound variable` error when calling the functions with empty arguments, or when attempting to get a value of an array element which had not been set previously. I could "fix" that as well but I think it is better as is.

For the first case (empty arguments), this can only happen when there is a mistake in the code you are using to call the function. I don't see anything wrong with the current behavior ('unbound variable' error). If you have a good argument against it, you are welcome to express it.

For the second case (getting a value of an unset array element), you will get the same error (unbound variable) with Bash arrays in a similar situation, and since my implementation attempts to follow Bash arrays behavior as much as possible, I think it's best to leave as is. Also this literally fulfills the purpose of the `-u` shell option.

An exception where you will **not** get the unbound variable error though you might expect one with the `-u` option is when trying to retrieve all values or all keys/indices of an array which had not been created. Rather you will get a 0 return code and an empty string as an output. This again follows the behavior of Bash arrays.

cataldor commented:
Thank you for the update friendly-bits! my comments inline:

> I updated the code for better compatibility with the -u shell option across all scripts. The `_check_vars()` function now performs a more thorough check of the arguments, which results in degraded performance, especially in the faster functions (`set_[x]_arr_el()` and `get_[x]_arr_val()`). To mitigate that, I implemented support for the `nocheckvars` environment variable which forces skipping these checks. It may be used in production after your code is well tested.

Excellent!
I also did not realize you could check the number of parameters received before calling `_check_args`.. this check is better than creating more variables in `_check_args` like i did in this PR.

> 
> You can still get an `unbound variable` error when calling the functions with empty arguments for array name or key/index, or when attempting to get a value of an array element which had not been set previously. I could "fix" that as well but I think it is better as is.

Personally, I agree with you. My concern was getting unbound variables in the "happy" path.

> 
> For the first case (empty arguments), this can only happen when there is a mistake in the code you are using to call the function. I don't see anything wrong with the current behavior ('unbound variable' error). If you have a good argument against it, you are welcome to express it.

Agreed. You should have passed the array name/key, so i think the unbound variable is working as expected here.

> 
> For the second case (getting a value of an unset array element), you will get the same error (unbound variable) with Bash arrays in a similar situation, and since my implementation attempts to follow Bash arrays behavior as much as possible, I think it's best to leave as is. Also this literally fulfills the purpose of the `-u` shell option.

Agreed.

> 
> An exception where you will **not** get the unbound variable error though you might expect one with the `-u` option is when trying to retrieve all values or all keys/indices of an array which had not been created. Rather you will get a 0 return code and an empty string as an output. This again follows the behavior of Bash arrays.

Agreed.

Before I close this PR, i do have one final question, but we may transfer this for an Issue if you prefer.
It is not clear to me how you are suppose to create the array for calling the functions in the mini versions, could you clarify that?

As you have seen, i have created copies of the declare arrays, but i'm not sure this is the expected behavior. 
Is it expected that you will always have the full version together with the mini?
Otherwise, it would be nice to clarify what a user has to do before calling the mini functions (set, get values).

friendly-bits commented:
> It is not clear to me how you are suppose to create the array for calling the functions in the mini versions, could you clarify that?

Maybe this is not prominent enough but it is in the README:

"The declare functions are not necessary to create an array. It's just a way to set N elements in one command. You can create an array by calling any function that stores a value in the array."

Does this answer your question?

friendly-bits commented:
> Is it expected that you will always have the full version together with the mini?

Not at all. The mini version can be used standalone. Moreover, the mini version is incompatible with the full version. Although you can technically add more functions from the full version to the mini version, these functions may need to be modified to work correctly.

cataldor commented:
> > It is not clear to me how you are suppose to create the array for calling the functions in the mini versions, could you clarify that?
> 
> Maybe this is not prominent enough but it is in the README:
> 
> "The declare functions are not necessary to create an array. It's just a way to set N elements in one command. You can create an array by calling any function that stores a value in the array."
> 
> The mini version only includes one function which stores a value in the array - set_[x]_arr_el().
> 
> Does this answer your question?

Yes!
My eyes passed right over this..

> 
> Edit: I updated the Usage section of the README to make this information more prominent.

GREAT!


> > Is it expected that you will always have the full version together with the mini?
> 
> Not at all. The mini version can be used standalone. Moreover, the mini version is incompatible with the full version. Although you can technically add more functions from the full version to the mini version, these functions may need to be modified to work correctly.

I understand.

Thank you! I will be closing this PR now
